### ISSUE TITLE
minor compaction never retries

### ISSUE THREAD

**Issue by ivakegg** [apache/accumulo#1644]
I came across a situation where a subset of minor compactions failed and were never retried (version 1.9.3).  This then resulted in a major compaction hanging in the TabletMemory.waitForMinC method forever.  Eventually the system backed with files on those tablets.
The reason the minor compaction thread died was because our system was being modified with a new context and the VFSClassLoader was changing classpaths.  This resulted in a NoClassDefFoundError while trying to load the iterators for that table.  The purpose of this ticket is not to fix anything to do with the VFSClassLoader mechanism, but rather to ensure that if a minor compaction fails for any reason, it will be subsequently retried perhaps with some sort of exponential backoff.

**Comment by milleruntime** [comment]
It looks like a NoClassDefFoundError will slip through the try/catch in MinorCompator [here](https://github.com/apache/accumulo/blob/0a9837f3f8395d89c5cd7bab7805c4aae28919be/server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/MinorCompactor.java#L136).   The easiest solution would be to add a NoClassDefFoundError to the catch next to UnsatisfiedLinkError.

**Comment by ivakegg** [comment]
I believe we need a little more than that.  If the minor compaction blows out for any reason, the TabletMemory should be tagged as failing such that the waitForMinC does not hang indefinitely, and perhaps also will throw an exception.

**Comment by ivakegg** [comment]
Perhaps the TabletServerResourcemanager.manageMemory should also do something when a minor compaction is detected to have failed......

**Comment by ivakegg** [comment]
> It looks like a NoClassDefFoundError will slip through the try/catch in MinorCompactor [here](https://github.com/apache/accumulo/blob/0a9837f3f8395d89c5cd7bab7805c4aae28919be/server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/MinorCompactor.java#L136). The easiest solution would be to add a NoClassDefFoundError to the catch next to UnsatisfiedLinkError.

The Comparactor.call method could also me modified to turn the NoClassDefFoundError into a RuntimeException which I believe gets handled appropriately in the MinorCompactor.call method.

**Comment by ivakegg** [comment]
The stack trace of the exception was something like this:
```
...
Caused by: java.lang.RuntimeException: Exception occurred during minor compaction on ...
   at org.apache.accumulo.tserver.tablet.Tablet.minorCompact
   at org.apache.accumulo.tserver.tablet.MinorCompactionTask.run
Caused by: java.lang.NoClassDefFoundError: Could not initialize class ...
   ...
   at org.apache.accumulo.core.iterators.IteratorUtil.loadIterators
   at org.apache.accumulo.tserver.tablet.Comparactor.compactLocalityGroup
   at org.apache.accumulo.tserver.tablet.Compactor.call
   at org.apache.accumulo.tserver.table.MinorCompactor.call
   at org.apache.accumulo.tserver.tablet.Tablet.minorCompact
```

**Comment by keith-turner** [comment]
> I believe we need a little more than that. If the minor compaction blows out for any reason, the TabletMemory should be tagged as failing such that the waitForMinC does not hang indefinitely, and perhaps also will throw an exception.

If you want to retry on everything then that would be all Errors and RuntimeExceptions, anything extending those two types  can be thrown without being declared.

```java
   } catch (RuntimeException | Error e) {
```

**Comment by keith-turner** [comment]
I think in #925 I chose to catch a specific error and retry because there are probably some errors that should not be retried.  The following javadoc has more info.  

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Error.html

Unfortunately it seems that we have run into two Errors where it completely makes sense to retry.  One worst case scenario for retrying on an error would be that it leads to corruption of persistent data that would not have occurred w/o the retry.

**Comment by ivakegg** [comment]
I think I agree we only want to add the NoClassDefFoundError specifically and handle that gracefully as we do for the RuntimeException.  I would however like to somehow mark the minor compaction as failed (in TabletMemory) for errors/exceptions that are not specifically caught and then have routines such as the waitForMinC to also throw an error in that scenario.

**Comment by keith-turner** [comment]
@ivakegg  one possible way to handle unexpected errors during minc would be for the tserver process to halt itself.

**Comment by ivakegg** [comment]
@keith-turner I am ok with halting the tserver.  Can I confirm that a failed minor compaction (which cannot be retried) is fatal?  @milleruntime was suggesting that this is only critical for the tablet and not the tserver.  In that case would we want to unload that tablet or (given the tserver is still standing) somehow continually generate an error message that we have a critical situation?

**Comment by brianloss** [comment]
Could these unexpected errors ever be due to a bad configuration? (e.g., Setting iterator properties manually or something like that?) I'd just be worried about halting the tablet server if it's not a transient problem since it will lead to cascading failure of the whole Accumulo cluster.

**Comment by keith-turner** [comment]
>  I'd just be worried about halting the tablet server if it's not a transient problem since it will lead to cascading failure of the whole Accumulo cluster.

That is a possibility.  However if unexpected errors cause minor compactions to hang indefinitely across the cluster, that will lead to another set of problems.  If there is something that automatically restarts tservers, halting may help work issues out automatically.

**Comment by keith-turner** [comment]
> @milleruntime was suggesting that this is only critical for the tablet and not the tserver. 

Many errors are a symptom of a wider problem.  For example when a thread tries to create another thread and gets an error, the problem is likely more widespread than the scope of the single thread.  Therefore some subset of error probably signify the entire tserver process is unhealthy in some way.

**Comment by keith-turner** [comment]
> Could these unexpected errors ever be due to a bad configuration? (e.g., Setting iterator properties manually or something like that?)

Hopefully in the case the iterator code throws a runtime exception, which I think are retried on when seen.

**Comment by brianloss** [comment]
> > I'd just be worried about halting the tablet server if it's not a transient problem since it will lead to cascading failure of the whole Accumulo cluster.
> 
> That is a possibility. However if unexpected errors cause minor compactions to hang indefinitely across the cluster, that will lead to another set of problems. If there is something that automatically restarts tservers, halting may help work issues out automatically.

Fair point. In a cluster with high load, restarting tablet servers is a fairly disruptive event. Having the tablet be unusable might be more disruptive, though. :) I missed that we're going to retry on the NoClassDefFound, so I agree for unknown errors it's better to halt the tablet server.

**Comment by ctubbsii** [comment]
This appears to have been fixed in #1646; please reopen if there is more to do.
